Fix SanitizeLabelName for certain edge case invalid labels

kelnage opened the pull request:
SanitizeLabelName does not correctly sanitize label names that:

1. Start with a digit (0-9)
2. Are empty

This commit changes the santization code to catch both of these edge cases and adds new tests to validate it works correctly in them both. In the first case, a leading digit will be replaced with an underscore, and in the latter, the function will return a single underscore.

@roidelapluie 

Signed-off-by: Nick Moore <nicholas.moore@grafana.com>

roidelapluie commented:
What issue is this fixing?

In Prometheus, this function is used to append label names after __meta labels, therefore it is fine to be empty or to start with a number. It would be a breaking change as is.

kelnage commented:
That makes sense - thanks for the context @roidelapluie. I think the issue is that the name of the function implies (at least it does to me) that it would sanitize _any_ label name, and other projects (e.g., Loki)  that depend on Prometheus appear to be using it in that way.

Would you prefer it if I instead:
1. renamed the originally implemented function to `SanitizePartialLabelName`, updated existing usage within the code, and kept my new implementation as is, or 
2. created a new function called `SanitizeFullLabelName` (and perhaps clarified the documentation for `SanitizeLabelName`)?

roidelapluie commented:
That sounds good to me

kelnage commented:
@roidelapluie I went with option 2 - hope that looks more reasonable. Not entirely sure why the tests are failing, given my changes should have minimal effect (if any) on the existing code base.

roidelapluie commented:
Thanks!
